text/BartekJachowicz/opis.tex

```diff
@@ -0,0 +1,129 @@
+\begin{lemma}
+Niech $h$ będzie największą liczbą taką że $A[i] = B[h]$ oraz $h < l_{i+1}(j - 1)$, wtedy liczba $L_i(j)$ jest definiowana przez $h$ oraz $L_{i+1}(j)$:
+$$
+y = \left\{ \begin{array}{ll}
+h & \textrm{gdy $ L_{i+1} $ jest zdefiniowane}\\
+L_{i+1}(j) & \textrm{gdy $ h $ nie istnieje}\\
+max(h, L_{i+1}(j) & \textrm{gdy obie wartości są zdefiniowane}
+\end{array} \right.
+$$
+\end{lemma}
+
+W procedurze wykorzystywana jest zmienna lokalna $lower_b$, która odpowiada za to, by nie przekroczyć wartości dla których $h$ oraz $L_{i+1}(j)$ z powyższego lematu dla danego $i, j$ są niezdefiniowane.\\
+Dodatkowo wprowadzona zostaje zmienna $over$, która odpowiada za to by przerwać działanie funkcji gdy wszystkie szukane wartości zostały już obliczone.
+
+Aby uzasadnić liniową złożoność pamięciową funkcji, wystarczy zauważyć że jedyne czego używa procedura \textit{fillone} to liniowej wielkości tablice $R_1$ oraz $R_2$, zatem złożoność pamięciowa to $O(n + m)$.
+Złożoność czasowa natomiast wynika z tego że maksymalna liczba porównań którą możemy wykonać w pętli \textit{while} to $n$, ponieważ nigdy nie zwiększamy zmiennej $pos_b$, zatem złożoność to $O(n)$.
+
+\subsection{Procedura \textit{calmid}}
+
+Procedura składa się z dwóch części. Pierwsza to pętla for, która w $i$-tej iteracji odpowiada za wyliczanie wartości $L_{m-i+2-j}(j)$ i zapisanie jej do $R_1[j]$. Przepisywanie wyników z tablicy $R_2$ do tablicy $R_1$ na koniec każdej iteracji zapewnia, że będą znajdowały się tam poprawne wartości dla wszystkich $0 \leq j \leq r$. Poprawność tych wartości wynika z poprawności działania procedury \textit{fillone} \\
+Druga część to przepisanie ostatecznych wartości $L_{m-x-j+1}(j)$ które zapisane są w tablicy $R_1(j)$ do zwracanej tablicy $LL(j)$.\\
+
+
+Podobnie jak przy procedurze \textit{fillone} złożoność pamięciowa oraz czasowa są dosyć proste do uzasadnienia.
+Wykorzystanie pamięci to ponownie jedynie linowego rozmiaru tablice $R_1, r_2, LL$, zatem złożoność pamięciowa to $O(n + m)$.\\
+Złożoność czasowa to natomiast $(x + 1)$ wywołanie procedury \textit{fillone} w pętli for. Przypominając że złożoność procedury \textit{fillone} wynosi $O(n)$, zatem złożoność czasowa procedury \textit{calmid} to $O((x+1)n)$.
+
+\subsection{Znajdowanie perfekcyjnego cięcia}
+
+Procedura znajdowania \textit{perfekcyjnego cięcia} polega na dwukrotnym wywołaniu procedury calmid odpowiednio dla odwróconych słów $A, B$ oraz dla nieodwróconych. Po zakończeniu tych wywołań otrzymujemy w tablicach $LL_1, LL_2$ wyliczone wartości odpowiednio: $n + 1 - L_{m-w+1-k}(k)$ dla $k \leq r_1$ oraz $L_{w+k+1}(m-(w+k)-w')$\ $(= L_{m-w'+1-(p-k)}(p-k))$ dla $p-k \leq r_2$. Z twierdzenia 1.1 wynika zatem, że wystarczy zatem dobrać $u = k + w$ oraz $v = LL_1(k)$ by dostać poprawne perfekcyjne cięcie.\\
```

Comment:
Maybe I'm missing something, but there is only mentioned L, but not L^*. Is this correct? If so, please explain why we may discard L^* even though it's mentioned in the theorem.

Reply:
I forgot to add lemma which explain the relation between $L_i(J)$ and $L_i^*(J)$